### ISSUE TITLE
Vulnerability: unpacked slackbuilds are owned by a random unprivileged user

### ISSUE THREAD

**Issue by pv4** [sbopkg/sbopkg#54]
Hello.

1.

check_asc() unpacks the slackbuild archive after checking gpg.

At the current time the files in slackbuild archives are owned by slackbuilds:slackbuilds (1016:1016). Since sbopkg is run as root these uid/gid are restored when unpacking and the unpacked files are owned by an unknown user 1016:1016 instead of root. If such a user exists he will have permissions to modify these files and inject arbitral commands that will be executed as root when the SlackBuild is run.

GNU tar has -o/--no-same-owner option to extract files as the current user (root). I think it must be used everywhere where a slackbuild archive is unpacked (not only in check_asc()).

2.

Another question (should I create another issue for it?) is whether it's necessary to rsync unpacked directories from slackbuilds.org mirror. They are not signed so we cannot verify their contents. Someone could hack the mirror and modify the unpacked directories.

Wouldn't it be more secure and more traffic-efficient to ignore directories and rsync only slackbuild archives and their signature files?

**Comment by willysr** [comment]
1. You mean --same-owner option?

2. sbopkg has a feature of editing the SlackBuild locally and this requires the original SlackBuild scripts available locally as well. Changes in the mirror (whether it's a legit or unauthorized) is not part of sbopkg's feature to check. It should be part of SBo's admin to make sure that won't happen.

**Comment by willysr** [comment]
2. You can also see that sbopkg has extract_tarball that extract the tarball again
# Re-extract the $APP tarball on top of local directory.  Can be used
# if tarball fails GPG check.

**Comment by pv4** [comment]
>     1. You mean --same-owner option?

No, exactly --no-same-owner:
```
$ tar --help |grep -C2 no-same-owner
                             cancel the effect of --delay-directory-restore
                             option
      --no-same-owner        extract files as yourself (default for ordinary
                             users)
      --no-same-permissions  apply the user's umask when extracting permissions
--

  -o                         when creating, same as --old-archive; when
                             extracting, same as --no-same-owner

 Other options:
```

> sbopkg has a feature of editing the SlackBuild locally

Could you provide more details please? If I edit SlackBuild locally, won't it be overwritten on the next build (sbopkg -b) in check_asc(), in when syncing next time (sbopkg -r)? If it's overwritten, what do you mean by "editing" if all my changes disappear before building/after syncing?

> Changes in the mirror (whether it's a legit or unauthorized) is not part of sbopkg's feature to check.

If it really were so, then sbopkg (and slackbuilds.org itself) would not have signature checking, simpler md5/shaXXX would be sufficient.

What is the reason to rsync these directories if we can just unpack them from rsync'ed archives which also would allow us to trust them? Is it performance?

**Comment by willysr** [comment]
If you run sbopkg (without any parameters), you will have an option to select a package and edit them locally and sbopkg will save it on a different filenames. You can either pick the locally-edited SlackBuild or use the regular SlackBuild from the repository.

sbopkg does check the tarball signature, but not for individual files in  the directory. The tarball is extracted during processing the SlackBuild for the that reason (if someone modified the individual files, it will be overwritten by the files extracted by the tarball). SBo guarantees the tarball's signature since it's always signed during each public update and the tarball was generated from the individual files. Those files were resides in public git repository where only limited people have access to it.

I guess it's just a simpler design to rsync all instead of selective sync. I'm not the one writing that feature. It was already there for years before i take over maintainership from original author.

**Comment by pv4** [comment]
> SBo guarantees the tarball's signature since it's always signed during each public update and the tarball was generated from the individual files.

I agree, and it does not guarantee the validness of these individual files. This makes using the files before extracting them from the verified archive insecure. Someone can hack the server and replace files in directories, and we are unable to verify whether they are correct. But if someone replaces archives he cannot also replace signatures because he does not have the private key, and gpg gives us the incorrect signature error.

> I guess it's just a simpler design to rsync all instead of selective sync. I'm not the one writing that feature. It was already there for years before i take over maintainership from original author.

However, changing this behavior would make sbopkg more secure but it also would require much testing and may lead to slower rsync because we would have to verify each individual slackbuild archive and then unpack it one by one.

But I think the first part of the issue (--no-same-owner) can be fixed easily, and I see no stoppers here.

**Comment by willysr** [comment]
> I agree, and it does not guarantee the validness of these individual files. This makes using the files before extracting them from the verified archive insecure. Someone can hack the server and replace files in directories, and we are unable to verify whether they are correct. But if someone replaces archives he cannot also replace signatures because he does not have the private key, and gpg gives us the incorrect signature error.

No, based on my understanding, the tarball is extracted over the local directory so if the individual files were modified, it will be overwritten during the extraction of the tarball, so it would still be secure.

I will do some testing on the --no-same-owner issue, but i can't promise on the time frame as i will be traveling in the next few days.

**Comment by pv4** [comment]
> the tarball is extracted over the local directory so if the individual files were modified, it will be overwritten during the extraction of the tarball, so it would still be secure.

You are correct. I meant that sbopkg guarantees that only AFTER tarball extraction (e.g. during check_asc()). Before that we just have what we got from slackbuilds.org and these files are not verified.

> I will do some testing on the --no-same-owner issue, but i can't promise on the time frame as i will be traveling in the next few days.

Great, thanks!

**Comment by willysr** [comment]
@pv4 : i wonder about this statement "At the current time the files in slackbuild archives are owned by slackbuilds:slackbuilds (1016:1016)."

Can you show your files owner/group?? On all of my machines, it's being owned by root:root

**Comment by pv4** [comment]
They are root:root right after rsync because `rsync` is called with `--no-owner` and `--no-group` flags:
```
root@vp:/var/lib/sbopkg/SBo/14.2/system# sbopkg -r
[...]
root@vp:/var/lib/sbopkg/SBo/14.2/system# ls -l mandoc
total 28
-rw-r--r-- 1 root root 1055 Jun 14  2018 README
-rw-r--r-- 1 root root  211 Jun 14  2018 configure-noconflict.local.gz
-rw-r--r-- 1 root root  205 Jun 14  2018 configure-replace.local.gz
-rw-r--r-- 1 root root  479 Jun 14  2018 man.conf.example
-rw-r--r-- 1 root root 3294 Jan 12 03:05 mandoc.SlackBuild
-rw-r--r-- 1 root root  270 Jan 12 03:05 mandoc.info
-rw-r--r-- 1 root root  848 Jun 14  2018 slack-desc
```
But the tarball is unpacked during build stage, and after that the files are owned by 1016:1016 (I guess because I have no slackbuilds:slackbuilds user):
```
root@vp:/var/lib/sbopkg/SBo/14.2/system# sbopkg -b mandoc
[...]
root@vp:/var/lib/sbopkg/SBo/14.2/system# ls -l mandoc
total 28
-rw-r--r-- 1 1016 1016 1055 Jun 14  2018 README
-rw-r--r-- 1 1016 1016  211 Jun 14  2018 configure-noconflict.local.gz
-rw-r--r-- 1 1016 1016  205 Jun 14  2018 configure-replace.local.gz
-rw-r--r-- 1 1016 1016  479 Jun 14  2018 man.conf.example
lrwxrwxrwx 1 root root   38 Mar 16 16:02 mandoc-1.14.4.tar.gz -> /var/cache/sbopkg/mandoc-1.14.4.tar.gz
-rwxr-xr-x 1 1016 1016 3294 Jan 12 03:05 mandoc.SlackBuild*
-rw-r--r-- 1 1016 1016  270 Jan 12 03:05 mandoc.info
-rw-r--r-- 1 1016 1016  848 Jun 14  2018 slack-desc
```
Also, you can view the slackbuild archives directly by `tar -tv` to see that the owner of the files is not root:
```
root@vp:/var/lib/sbopkg/SBo/14.2/system# sbopkg -v
0.38.1
root@vp:/var/lib/sbopkg/SBo/14.2/system# tar -tvf mandoc.tar.gz
drwxr-xr-x slackbuilds/slackbuilds 0 2019-01-12 03:05 mandoc/
-rw-r--r-- slackbuilds/slackbuilds 211 2018-06-14 17:01 mandoc/configure-noconflict.local.gz
-rw-r--r-- slackbuilds/slackbuilds 270 2019-01-12 03:05 mandoc/mandoc.info
-rwxr-xr-x slackbuilds/slackbuilds 3294 2019-01-12 03:05 mandoc/mandoc.SlackBuild
-rw-r--r-- slackbuilds/slackbuilds  848 2018-06-14 17:01 mandoc/slack-desc
-rw-r--r-- slackbuilds/slackbuilds 1055 2018-06-14 17:01 mandoc/README
-rw-r--r-- slackbuilds/slackbuilds  205 2018-06-14 17:01 mandoc/configure-replace.local.gz
-rw-r--r-- slackbuilds/slackbuilds  479 2018-06-14 17:01 mandoc/man.conf.example
```

Btw, I have been using sbopkg for the last week with the modification I propose (the patch attached) and I haven't had any issues (during this time I built 2-3 packages using sbopkg).

[sbopkg-0.38.1.diff.gz](https://github.com/sbopkg/sbopkg/files/2973926/sbopkg-0.38.1.diff.gz)
